beetsplug/lyrics.py

```diff
@@ -209,105 +193,151 @@ def slug(text):
+class RequestHandler:
+    _log: beets.logging.Logger
 
-    def try_parse_html(html, **kwargs):
-        return bs4.BeautifulSoup(html, "html.parser", **kwargs)
+    def debug(self, message: str, *args: Any) -> None:
+        """Log a debug message with the class name."""
+        self._log.debug(f"{self.__class__.__name__}: {message}", *args)
 
-else:
+    def info(self, message: str, *args: Any) -> None:
+        """Log an info message with the class name."""
+        self._log.info(f"{self.__class__.__name__}: {message}", *args)
 
-    def try_parse_html(html, **kwargs):
-        return None
+    def warn(self, message: str, *args: Any) -> None:
+        """Log warning with the class name."""
+        self._log.warning(f"{self.__class__.__name__}: {message}", *args)
+
+    def fetch_text(self, url: str, **kwargs) -> str:
+        """Return text / HTML data from the given URL."""
+        self.debug("Fetching HTML from {}", url)
+        return r_session.get(url, **kwargs).text
+
+    def fetch_json(self, url: str, **kwargs):
+        """Return JSON data from the given URL."""
+        self.debug("Fetching JSON from {}", url)
+        return r_session.get(url, **kwargs).json()
+
+    @contextmanager
+    def handle_request(self) -> Iterator[None]:
+        try:
+            yield
+        except requests.JSONDecodeError:
+            self.warn("Could not decode response JSON data")
+        except requests.RequestException as exc:
+            self.warn("Request error: {}", exc)
+
+
+class BackendType(type):
+    """Metaclass for the :class:`Backend` class.
+
+    It keeps track of defined subclasses and provides access to them through
+    the base class:
+
+    >>> Backend["genius"]  # beetsplug.lyrics.Genius
+    >>> list(Backend)  # ["lrclib", "musixmatch", "genius", "tekstowo", "google"]
+    """
 
+    _registry: dict[str, BackendType] = {}
+    REQUIRES_BS: bool
 
-class Backend:
+    def __new__(cls, name: str, bases: tuple[type, ...], attrs) -> BackendType:
+        """Create a new instance of the class and add it to the registry."""
+        new_class = super().__new__(cls, name, bases, attrs)
+        if bases:
+            cls._registry[name.lower()] = new_class
+        return new_class
+
+    @classmethod
+    def __getitem__(cls, key: str) -> BackendType:
+        return cls._registry[key]
+
+    @classmethod
+    def __iter__(cls) -> Iterator[str]:
+        return iter(cls._registry)
+
+
+class Backend(RequestHandler, metaclass=BackendType):
     REQUIRES_BS = False
 
     def __init__(self, config, log):
         self._log = log
         self.config = config
 
-    @staticmethod
-    def _encode(s):
-        """Encode the string for inclusion in a URL"""
-        if isinstance(s, str):
-            for char, repl in URL_CHARACTERS.items():
-                s = s.replace(char, repl)
-            s = s.encode("utf-8", "ignore")
-        return urllib.parse.quote(s)
+    def fetch(self, artist, title, album=None, length=None):
```

Comment:
This function should have a clear type signature, and it needs more documentation.  Are the parameters matched fuzzily or exactly?

Reply:
I ignored it and left the way it looked like previously, though I can surely add types here

Reply:
Simplified this method to accept `item` in the last commit